File: sht1/argonfluid/amclj.c
/**********************************************************************
 *
 * File: amclj.c
 *
 * NVT-Monte Carlo of Lennard-Jonesium
 * Analyze checkpoint file
 *
 * 12-May-2007 (MN)
 * 19-Apr-2012
 *
 **********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "getval.h"

#define NL '\n'
#define BSIZE 80

/**********************************************************************/

int main() {

/**********************************************************************/

  const double blue=0.0,green=0.0,radius=0.5,red=1.0;

  FILE *fpi,*fpo;
  char copy;
  char fname[BSIZE];
  unsigned short seed[3];
  int n,nt,ntjob,ntprint,ntskip;
  int i,ndr;
  double disp,dr,rho,t;
  double c,c2,cv,fact,p,r,r1,r2;
  double *x,*y,*z;
  long long *ag;
  double accr,au,au2,aw;

  /* Read checkpoint file */

  strcpy(fname,"mclj_out.dat");
  printf(" fname=[mclj_out.dat] ");
  getval_s(fname);

  fpi=fopen(fname,"r");

  fread(&n,sizeof(int),1,fpi);
  fread(&nt,sizeof(int),1,fpi);
  fread(&ntjob,sizeof(int),1,fpi);
  fread(&ntprint,sizeof(int),1,fpi);
  fread(&ntskip,sizeof(int),1,fpi);
  fread(seed,sizeof(unsigned short),3,fpi);

  /* Check for zero-length run */

  if(nt<=0) {
    fclose(fpi);
    printf(" amclj: empty file\n");
    exit(0);
  }

  /* Simulation parameters */

  fread(&disp,sizeof(double),1,fpi);
  fread(&dr,sizeof(double),1,fpi);
  fread(&rho,sizeof(double),1,fpi);
  fread(&t,sizeof(double),1,fpi);

  /* Allocate arrays */

  c=cbrt(n/rho);
  c2=0.5*c;
  ndr=c2/dr;

  ag=(long long*)malloc(ndr*sizeof(long long));
  x=(double*)malloc(n*sizeof(double));
  y=(double*)malloc(n*sizeof(double));
  z=(double*)malloc(n*sizeof(double));

  /* Positions, accumulated averages & g(r) */

  fread(x,sizeof(double),n,fpi);
  fread(y,sizeof(double),n,fpi);
  fread(z,sizeof(double),n,fpi);

  fread(&accr,sizeof(double),1,fpi);
  fread(&au,sizeof(double),1,fpi);
  fread(&au2,sizeof(double),1,fpi);
  fread(&aw,sizeof(double),1,fpi);

  fread(ag,sizeof(long long),ndr,fpi);

  fclose(fpi);

  /* Print results: simulation parameters */

  printf("\n");
  printf("     n=%9d\n",n);
  printf("   rho=%9.5lf\n",rho);
  printf("     t=%9.5lf\n",t);
  printf("  disp=%9.5lf\n",disp);
  printf("\n");

  /* Averages */

  cv=1.5+n*(au2/nt-(au/nt)*(au/nt))/(t*t);
  p=rho*(t-aw/(3*nt));
  printf("    nt=%12d (*%5d)\n",nt,ntskip);
  printf("  accr=%12.5e\n",accr/nt);
  printf(" <U>/N=%12.5e\n",au/nt);
  printf("  Cv/N=%12.5e\n",cv);
  printf("     p=%12.5e\n",p);
  printf("\n");

  /* Write g(r)? */

  printf(" Write g(r) to 'amclj.dat? [y] ");
  copy=(char)fgetc(stdin);

  if(copy=='y'||copy=='Y'||copy==NL) {

    fact=3.0/(2.0*M_PI*rho*n*nt);
    fpo=fopen("amclj.dat","w");

    for(i=0;i<=ndr-2;i++) {
      r1=i*dr;
      r2=(i+1)*dr;
      r=0.5*(r1+r2);
      fprintf(fpo," %8.5lf %12.5e\n",\
	      r,fact*ag[i]/(r2*r2*r2-r1*r1*r1));
    }

    fclose(fpo);

  }

  if(copy!=NL) {
    copy=(char)fgetc(stdin);
  }
  
  /* Write PDB file? */

  printf(" Write PDB format to amclj.pdb? [y] ");
  copy=(char)fgetc(stdin);

  if(copy=='y'||copy=='Y'||copy==NL) {

    fpo=fopen("amclj.pdb","w");

    fprintf(fpo,"CRYST1"" %8.3lf %8.3lf %8.3lf",c,c,c);
    fprintf(fpo," %6.2f %6.2f %6.2f",90.0,90.0,90.0);
    fprintf(fpo,"  P1           1\n");

    for(i=0;i<=n-1;i++) {
      fprintf(fpo,"HETATM%5d                    %7.3lf %7.3lf %7.3lf\n",\
	      i+1,x[i]+c2,y[i]+c2,z[i]+c2);
    }

    fprintf(fpo,"COLOR ##### ####              ");
    fprintf(fpo," %7.3lf %7.3lf %7.3lf %5.2lf\n",red,green,blue,radius);
    fprintf(fpo,"END\n");

    fclose(fpo);

  }

  /* Deallocate arrays */

  free(ag);
  free(x);
  free(y);
  free(z);

  return 0;

}

/**********************************************************************/

File: sht1/argonfluid/fortran/amclj.f90
!**********************************************************************!
!
! File: amclj.f90
!
! NVT-Monte Carlo of Lennard-Jonesium
! Analyze checkpoint file
!
! 13-Apr-1997 (MN)
! 19-Apr-2012
!
!**********************************************************************!

program amclj

!**********************************************************************!

  implicit none

  character(len=1)::copy
  integer,parameter::long=selected_int_kind(18)
  integer,parameter::double=selected_real_kind(15)
  integer,parameter::iin=1,iout=2
  real,parameter::blue=0.0,green=0.0,radius=0.5,red=1.0

  character(len=80)::fname
  integer::i,n,ndr,nt,ntjob,ntprint,ntskip
  real::c,c2,cv,disp,dr,fact,p,pi,r,r1,r2,rho,t
  real,dimension(:),allocatable::x,y,z
  integer(kind=long),dimension(:),allocatable::ag
  real(kind=double)::accr,au,au2,aw

! Read checkpoint file

  write(unit=*,fmt="(a)",advance="no") " fname=[mclj_out.dat] "
  read(unit=*,fmt="(a)") fname
  if(fname=="") then
    fname="mclj_out.dat"
  end if

  open(unit=iin,file=fname,status="old",action="read", &
       form="unformatted",position="rewind")

  read(unit=iin) n,nt,ntjob,ntprint,ntskip!,iran

! Check for zero-length run

  if(nt<=0) then
    write(unit=*,fmt="(a)") " amclj: empty file"
    close(unit=iin)
    stop
  end if

! Simulation parameters

  read(unit=iin) disp,dr,rho,t

! Allocate arrays

  c=(n/rho)**(1.0/3.0)
  c2=0.5*c
  ndr=int(c2/dr)

  allocate(ag(0:ndr-1),x(n),y(n),z(n))

! Positions, accumulated averages & g(r)

  read(unit=iin) x,y,z
  read(unit=iin) accr,au,au2,aw
  read(unit=iin) ag

  close(unit=iin)

! Suppress compiler warnings

  ntjob=ntjob
  ntprint=ntprint

! Print results: simulation parameters

  write(unit=*,fmt="()")
  write(unit=*,fmt="(a,i9)") "     n=",n
  write(unit=*,fmt="(a,f9.5)") "   rho=",rho
  write(unit=*,fmt="(a,f9.5)") "     t=",t
  write(unit=*,fmt="(a,f9.5)") "  disp=",disp
  write(unit=*,fmt="()")

! Averages

  cv=real(1.5+n*(au2/nt-(au/nt)**2)/t**2)
  p=real(rho*(t-aw/(3*nt)))

  write(unit=*,fmt="(a,i12,a,i5,a)") "    nt=",nt," (*",ntskip,")"
  write(unit=*,fmt="(a,es12.5)") "  accr=",accr/nt
  write(unit=*,fmt="(a,es12.5)") " <U>/N=",au/nt
  write(unit=*,fmt="(a,es12.5)") "  Cv/N=",cv
  write(unit=*,fmt="(a,es12.5)") "     p=",p
  write(unit=*,fmt="()")

! Write g(r) to file?

  write(unit=*,fmt="(a)",advance="no") &
       " Write g(r) to 'amclj.dat'? [y] "
  read(unit=*,fmt="(a)") copy

  if(copy=="y".or.copy=="Y".or.copy=="") then

    pi=4.0*atan(1.0)
    fact=3.0/(2.0*pi*rho*n*nt)

    open(unit=iout,file="amclj.dat",status="replace",action="write", &
         form="formatted")

    do i=0,ndr-1
      r1=i*dr
      r2=(i+1)*dr
      r=0.5*(r1+r2)
      write(unit=iout,fmt="(tr1,f8.5,tr1,es12.5)") &
           r,fact*ag(i)/(r2**3-r1**3)
    end do

    close(unit=iout)

  end if

! Write PDB file?

  write(unit=*,fmt="(a)",advance="no") &
       " Write PDB format to 'amclj.pdb'? [y] "
  read(unit=*,fmt="(a)") copy

  if(copy=="y".or.copy=="Y".or.copy=="") then

    open(unit=iout,file="amclj.pdb",status="replace",action="write", &
         form="formatted")

    write(unit=iout,fmt="(a,3f9.3,3f7.2,tr2,a,tr11,a)") &
         "CRYST1",c,c,c,90.0,90.0,90.0,"P1","1"
    do i=1,n
      write(unit=iout,fmt="(a,i5,tr19,3f8.3)") &
           "HETATM",i,x(i)+c2,y(i)+c2,z(i)+c2
    end do
    write(unit=iout,fmt= &
         "(a,a,tr1,a,tr14,3f8.3,f6.2)") &
         "COLOR ","#####","####",red,green,blue,radius
    write(unit=iout,fmt="(a)") "END"

    close(unit=iout)

  end if

! Deallocate arrays

  deallocate(ag,x,y,z)

end program amclj

!**********************************************************************!
